crates/relayer/src/chain/ckb4ibc/monitor.rs

```diff
@@ -0,0 +1,357 @@
+                        channel_id: Some(item.0.channel_id),
+                        connection_id: item.0.channel_end.connection_hops[0].clone(),
+                        counterparty_port_id: item.0.channel_end.remote.port_id,
+                        counterparty_channel_id: item.0.channel_end.remote.channel_id,
+                    }),
+                    height: Height::new(1, 1).unwrap(), // todo
+                    tx_hash: item.1.into(),
+                },
+                _ => unreachable!(),
+            })
+            .collect::<Vec<_>>();
+        Ok(EventBatch {
+            chain_id: self.config.id.clone(),
+            tracking_id: TrackingId::Static("ckb channel events collection"),
+            height: Height::new(1, 1).unwrap(), // todo
+            events,
+        })
+    }
+
+    async fn fetch_packet_events(&self) -> Result<EventBatch> {
+        let key = SearchKey {
+            script: Script::new_builder()
+                .code_hash(self.config.packet_code_hash.pack())
+                .args("".pack())
+                .build()
+                .into(),
+            script_type: ScriptType::Lock,
+            filter: None,
+            with_data: None,
+            group_by_transaction: None,
+        };
+        let ibc_packets = self
+            .search_and_extract(
+                key,
+                &|tx| {
+                    let hash = tx.hash.clone();
+                    let obj = extract_ibc_packet_from_tx(tx)
+                        .map_err(|_| Error::collect_events_failed("packet".to_string()))?;
+                    Ok((obj, hash))
+                },
+                20,
+            )
+            .await?;
+        let events = ibc_packets
+            .into_iter()
+            .map(|item| match item.0.status {
+                PacketStatus::Send => IbcEventWithHeight {
+                    event: IbcEvent::SendPacket(SendPacket {
+                        packet: convert_packet(item.0),
+                    }),
+                    height: Height::new(1, 1).unwrap(), // todo
+                    tx_hash: item.1.into(),
+                },
+                PacketStatus::Recv => IbcEventWithHeight {
+                    event: IbcEvent::ReceivePacket(ReceivePacket {
+                        packet: convert_packet(item.0),
+                    }),
+                    height: Height::new(1, 1).unwrap(), // todo
+                    tx_hash: item.1.into(),
+                },
+                PacketStatus::Ack => todo!(),
+                PacketStatus::InboxAck => todo!(),
+                PacketStatus::OutboxAck => todo!(),
+            })
+            .collect::<Vec<_>>();
+        Ok(EventBatch {
+            chain_id: self.config.id.clone(),
+            tracking_id: TrackingId::Static("ckb channel events collection"),
+            height: Height::new(1, 1).unwrap(), // todo
+            events,
+        })
+    }
+
+    async fn search_and_extract<T, F>(
```

Comment:
I'm not sure whether we should cache the searched results about request cells for connection, channel, and packet, because the `search_and_extract` is running in a loop, so emitting the same ibc events might be possible I guess

If processing duplicated ibc events is fine, it's ok, but if not, we have to do something to prevent it